sherpa/astro/utils/tests/test_astro_utils_unit.py

```diff
@@ -220,3 +224,269 @@ def test_do_group_single_group(func, expected):
+    assert utils.calc_data_sum(data, *frange) == expected
+    assert not data.mask
+
+
+@pytest.mark.parametrize("frange,expected", [((0, 20), 14), ((0, 10), 9),
+                                             ((2, 3), 2),
+                                             ((2, 10), 9),((3, 7), 3),
+                                             ((4, 6), 3),
+                                             ((3, 12), 7), ((1, 13), 14),
+                                             ((20, 22), 0)])
+def test_calc_data_sum_filtered_1dint(frange, expected):
+    """Filter out everything and then check with both lo/hi set: Data1DInt"""
+
+    data = make_data("1dint")
+    data.ignore()
+    assert not data.mask
+    assert utils.calc_data_sum(data, *frange) == expected
+    assert not data.mask
+
+
+@pytest.mark.parametrize("frange,expected", [((0, 10), 14), ((0, 4), 14),
+                                             ((2, 3), 7),
+                                             ((2, 10), 12),
+                                             ((1, 1), 2),
+                                             ((2, 2), 3),
+                                             ((3, 3), 4),
+                                             ((4, 4), 5),
+                                             ((3, 4), 9),
+                                             ((3, 5), 9),
+                                             ((3, 12), 9),
+                                             ((1, 13), 14),
+                                             ((20, 22), 0)])
+def test_calc_data_sum_filtered_pha(frange, expected):
+    """Filter out everything and then check with both lo/hi set: DataPHA"""
+
+    data = make_data("pha")
+    data.ignore()
+    assert not data.mask
+    assert utils.calc_data_sum(data, *frange) == expected
+    assert not data.mask
+
+
+@pytest.mark.parametrize("data_class", ["grp", "qual"])
+@pytest.mark.parametrize("frange,expected", [((0, 10), 14), ((0, 4), 14),
+                                             ((2, 3), 9),
+                                             ((2, 10), 14),
+                                             ((1, 1), 5),
+                                             ((2, 2), 5),
+                                             ((3, 3), 4),  # this is technically filtered-out by bad quality
+                                             ((4, 4), 5),
+                                             ((3, 4), 9),
+                                             ((3, 5), 9),
+                                             ((3, 12), 9),
+                                             ((1, 13), 14),
+                                             ((20, 22), 0)])
+def test_calc_data_sum_filtered_pha_grouped(frange, expected, data_class):
+    """Filter out everything and then check with both lo/hi set: DataPHA (grouped/quality)"""
+
+    data = make_data(data_class)
+    data.ignore()
+    assert not data.mask
+    assert utils.calc_data_sum(data, *frange) == expected
```

Comment:
The longer I look at this, the less I understand what `calc_data_sum` is good for.

Reply:
It actually comes down to the work I did for the 4.14 release to try and "harmonise" the range filtering of ignore/notice. The issue is whether a filter of  `lo,hi` is `[lo, hi]` (inclusive for both the `lo` and `hi` values) - which we have for `Data1D`, or `[lo, hi)` (so exclusive for the `hi` limit) as have do for `Data1DInt`. `DataPHA` is a special case because technically we have `[lo,hi]` for channel and `[lo,hi)` for energy or wavelength filtering [*], but these particular tests only really test the `channel` case (since we rely on the existing `notice` tests to properly check all the many variants of this).

[*] if memory serves, we did actually discuss this in your reviews of the filter code, but it is easy to get confused.

To go off on a bit of a rant - I think the original `calc_data_sum` code assumes we are using energies and hasn't really thought through the cases of using wavelength or channel, or with a non-PHA data set. I am not willing to put a lot of effort into improving this at this time.

Reply:
That's fine - I'm not asking to improve the meaning of the method. Maybe just your magic "this is a regression test, there may be issues with calculating the sum in channel or wavelength"?